# Importing Libraries

In [7]:
import numpy as np
import pandas as pd

# Loading Dataset

In [8]:
job_df = pd.read_csv(r"C:\Users\vnav2\OneDrive\Desktop\Combined_Jobs_Final.csv",header=0)

In [9]:
job_df

Job.ID  Provider Status  \
0         111         1   open   
1         113         1   open   
2         117         1   open   
3         121         1   open   
4         127         1   open   
...       ...       ...    ...   
84085      82         1   open   
84086      83         1   open   
84087      84         1   open   
84088      88         1   open   
84089      92         1   open   

                                                    Slug  \
0                        palo-alto-ca-tacolicious-server   
1        san-francisco-ca-claude-lane-kitchen-staff-chef   
2      san-francisco-ca-machka-restaurants-corp-barte...   
3                      brisbane-ca-teriyaki-house-server   
4      los-angeles-ca-rosa-mexicano-sunset-kitchen-st...   
...                                                  ...   
84085  san-francisco-ca-national-japanese-american-hi...   
84086       larkspur-ca-emporio-rulli-kitchen-staff-chef   
84087                san-francisco-ca-onigilly-driver-84   
84088  san-francisco-ca-machka-restaurants-corp-line-...   
84089               san-jose-ca-kazoo-restaurant-cashier   

                                                   Title            Position  \
0                                   Server @ Tacolicious              Server   
1                       Kitchen Staff/Chef @ Claude Lane  Kitchen Staff/Chef   
2                   Bartender @ Machka Restaurants Corp.           Bartender   
3                                Server @ Teriyaki House              Server   
4            Kitchen Staff/Chef @ Rosa Mexicano - Sunset  Kitchen Staff/Chef   
...                                                  ...                 ...   
84085  Book Keeper @ National Japanese American Histo...         Book Keeper   
84086                 Kitchen Staff/Chef @ Emporio Rulli  Kitchen Staff/Chef   
84087                                  Driver @ Onigilly              Driver   
84088               Line Cook @ Machka Restaurants Corp.           Line Cook   
84089                         Cashier @ Kazoo Restaurant             Cashier   

                                             Company           City  \
0                                        Tacolicious      Palo Alto   
1                                        Claude Lane  San Francisco   
2                           Machka Restaurants Corp.  San Francisco   
3                                     Teriyaki House       Brisbane   
4                             Rosa Mexicano - Sunset    Los Angeles   
...                                              ...            ...   
84085  National Japanese American Historical Society  San Francisco   
84086                                  Emporio Rulli       Larkspur   
84087                                       Onigilly  San Francisco   
84088                       Machka Restaurants Corp.  San Francisco   
84089                               Kazoo Restaurant       San Jose   

       State.Name State.Code  ...               Industry  \
0      California         CA  ...     Food and Beverages   
1      California         CA  ...     Food and Beverages   
2      California         CA  ...     Food and Beverages   
3      California         CA  ...     Food and Beverages   
4      California         CA  ...     Food and Beverages   
...           ...        ...  ...                    ...   
84085  California         CA  ...  Office Administration   
84086  California         CA  ...     Food and Beverages   
84087  California         CA  ...     Food and Beverages   
84088  California         CA  ...     Food and Beverages   
84089  California         CA  ...     Food and Beverages   

                                         Job.Description  Requirements Salary  \
0      Tacolicious' first Palo Alto store just opened...           NaN   8.00   
1       \r\n\r\nNew French Brasserie in S.F. Financia...           NaN   0.00   
2      We are a popular Mediterranean wine bar and re...           NaN  11.00   
3       ● Serve fo

# The columns  describe various attributes related to a job listing or job posting. Here's an explanation of each one:

1. Job.ID: A unique identifier for each job listing. This can be a numeric or alphanumeric code used to distinguish each job post.

2. Provider: The entity or platform that is providing or hosting the job listing. This could refer to the job board, company, or agency that posted the job.

3. Status: Indicates the current status of the job listing, such as "open," "closed," "filled," or "pending."

4. Slug: A URL-friendly version of the job title or description, typically used for web pages. It's usually lowercase and contains hyphens instead of spaces (e.g., data-analyst-new-york).

5. Title: The job title or position name, such as "Data Scientist," "Software Engineer," etc.

6. Position: This could refer to the level or type of the job position, such as "Junior," "Senior," "Intern," or "Manager."

7. Company: The company or organization offering the job position.

8. City: The city where the job is located.

9. State.Name: The name of the state or province where the job is located.

10. State.Code: The abbreviated code for the state or province (e.g., "CA" for California, "NY" for New York).

11. Industry: The industry or sector to which the job belongs, such as "Technology," "Healthcare," "Finance," etc.

12. Job.Description: A detailed description of the job duties, responsibilities, and expectations.

13. Requirements: The qualifications or skills required for the job, such as education, experience, certifications, or technical skills.

14. Salary: The compensation offered for the job position, which may include an hourly wage, annual salary, or bonus information.

15. Listing.Start: The date when the job listing was posted or became active.

16. Listing.End: The date when the job listing expires or is no longer available.

17. Employment.Type: The type of employment, such as "Full-time," "Part-time," "Contract," or "Temporary."

18. Education.Required: The minimum level of education required for the job, such as "Bachelor's Degree," "Master's Degree," or "High School Diploma."

19. Created.At: The timestamp indicating when the job listing was first created.

20. Updated.At: The timestamp indicating when the job listing was last updated or modified.

These columns together provide a comprehensive overview of each job listing, including its details, location, employer, requirements, and other attributes related to the job.

In [10]:
job_df.head(3)

Job.ID  Provider Status                                               Slug  \
0     111         1   open                    palo-alto-ca-tacolicious-server   
1     113         1   open    san-francisco-ca-claude-lane-kitchen-staff-chef   
2     117         1   open  san-francisco-ca-machka-restaurants-corp-barte...   

                                  Title            Position  \
0                  Server @ Tacolicious              Server   
1      Kitchen Staff/Chef @ Claude Lane  Kitchen Staff/Chef   
2  Bartender @ Machka Restaurants Corp.           Bartender   

                    Company           City  State.Name State.Code  ...  \
0               Tacolicious      Palo Alto  California         CA  ...   
1               Claude Lane  San Francisco  California         CA  ...   
2  Machka Restaurants Corp.  San Francisco  California         CA  ...   

             Industry                                    Job.Description  \
0  Food and Beverages  Tacolicious' first Palo Alto store just opened...   
1  Food and Beverages   \r\n\r\nNew French Brasserie in S.F. Financia...   
2  Food and Beverages  We are a popular Mediterranean wine bar and re...   

   Requirements Salary Listing.Start  Listing.End  Employment.Type  \
0           NaN    8.0           NaN          NaN        Part-Time   
1           NaN    0.0           NaN          NaN        Part-Time   
2           NaN   11.0           NaN          NaN        Part-Time   

  Education.Required               Created.At               Updated.At  
0                NaN  2013-03-12 02:08:28 UTC  2014-08-16 15:35:36 UTC  
1                NaN  2013-04-12 08:36:36 UTC  2014-08-16 15:35:36 UTC  
2                NaN  2013-07-16 09:34:10 UTC  2014-08-16 15:35:37 UTC  

[3 rows x 23 columns]

In [59]:
job_df.info

<bound method DataFrame.info of       Status                                              Title  \
64119   open                     site director knowledg univers   
35827   open                        administr assist officeteam   
72100   open                   account manag chi payment system   
46355   open  outsid wholesal sale rep parttim river front c...   
34166   open  custom servic rep help peopl hear loss captioncal   
...      ...                                                ...   
66282   open  sale repres sale associ entri level vector market   
39515   open                           staff account accountemp   
69231   open                  unarm secur offic us secur associ   
69618   open            line cook crown plaza independ own oper   
6144    open  kitchen manag job detail artesian hotel casino...   

                                                Position  \
64119                                      Site Director   
35827                           Administrat

In [12]:
job_df.tail(3)

Job.ID  Provider Status  \
84087      84         1   open   
84088      88         1   open   
84089      92         1   open   

                                                    Slug  \
84087                san-francisco-ca-onigilly-driver-84   
84088  san-francisco-ca-machka-restaurants-corp-line-...   
84089               san-jose-ca-kazoo-restaurant-cashier   

                                      Title   Position  \
84087                     Driver @ Onigilly     Driver   
84088  Line Cook @ Machka Restaurants Corp.  Line Cook   
84089            Cashier @ Kazoo Restaurant    Cashier   

                        Company           City  State.Name State.Code  ...  \
84087                  Onigilly  San Francisco  California         CA  ...   
84088  Machka Restaurants Corp.  San Francisco  California         CA  ...   
84089          Kazoo Restaurant       San Jose  California         CA  ...   

                 Industry                                    Job.Description  \
84087  Food and Beverages  ONIGILLY (Japanese rice ball wraps) seeks outg...   
84088  Food and Beverages  We are a popular Mediterranean restaurant in F...   
84089  Food and Beverages   We are looking for a cashier! \r\n\r\n ● Take...   

       Requirements Salary Listing.Start  Listing.End  Employment.Type  \
84087           NaN   11.0           NaN          NaN        Part-Time   
84088           NaN   13.0           NaN          NaN        Part-Time   
84089           NaN   10.0           NaN          NaN        Part-Time   

      Education.Required               Created.At               Updated.At  
84087                NaN  2013-03-12 01:47:13 UTC  2014-08-16 15:35:27 UTC  
84088                NaN  2013-07-16 08:55:22 UTC  2014-08-16 15:35:28 UTC  
84089                NaN  2013-03-27 09:35:04 UTC  2014-08-16 15:35:30 UTC  

[3 rows x 23 columns]

In [13]:
job_df[['Title','Job.Description']]

Title  \
0                                   Server @ Tacolicious   
1                       Kitchen Staff/Chef @ Claude Lane   
2                   Bartender @ Machka Restaurants Corp.   
3                                Server @ Teriyaki House   
4            Kitchen Staff/Chef @ Rosa Mexicano - Sunset   
...                                                  ...   
84085  Book Keeper @ National Japanese American Histo...   
84086                 Kitchen Staff/Chef @ Emporio Rulli   
84087                                  Driver @ Onigilly   
84088               Line Cook @ Machka Restaurants Corp.   
84089                         Cashier @ Kazoo Restaurant   

                                         Job.Description  
0      Tacolicious' first Palo Alto store just opened...  
1       \r\n\r\nNew French Brasserie in S.F. Financia...  
2      We are a popular Mediterranean wine bar and re...  
3       ● Serve food/drinks to customers in a profess...  
4      Located at the heart of Hollywood, we are one ...  
...                                                  ...  
84085  NJAHS stands for National Japanese American Hi...  
84086  Weekend Brunch Line Cook \r\n● Other shifts ma...  
84087  ONIGILLY (Japanese rice ball wraps) seeks outg...  
84088  We are a popular Mediterranean restaurant in F...  
84089   We are looking for a cashier! \r\n\r\n ● Take...  

[84090 rows x 2 columns]

### job_df: This is your DataFrame, which likely contains various job-related information.

[['Title', 'Job.Description']]: This is a list of column names you want to extract from the DataFrame. In this case, the selected columns are:

'Title': Likely refers to the job title (e.g., Software Engineer, Data Scientist, etc.).
'Job.Description': Likely contains the job description text detailing the responsibilities, qualifications, and other aspects of the job.

In [14]:
job_df.head(3)

Job.ID  Provider Status                                               Slug  \
0     111         1   open                    palo-alto-ca-tacolicious-server   
1     113         1   open    san-francisco-ca-claude-lane-kitchen-staff-chef   
2     117         1   open  san-francisco-ca-machka-restaurants-corp-barte...   

                                  Title            Position  \
0                  Server @ Tacolicious              Server   
1      Kitchen Staff/Chef @ Claude Lane  Kitchen Staff/Chef   
2  Bartender @ Machka Restaurants Corp.           Bartender   

                    Company           City  State.Name State.Code  ...  \
0               Tacolicious      Palo Alto  California         CA  ...   
1               Claude Lane  San Francisco  California         CA  ...   
2  Machka Restaurants Corp.  San Francisco  California         CA  ...   

             Industry                                    Job.Description  \
0  Food and Beverages  Tacolicious' first Palo Alto store just opened...   
1  Food and Beverages   \r\n\r\nNew French Brasserie in S.F. Financia...   
2  Food and Beverages  We are a popular Mediterranean wine bar and re...   

   Requirements Salary Listing.Start  Listing.End  Employment.Type  \
0           NaN    8.0           NaN          NaN        Part-Time   
1           NaN    0.0           NaN          NaN        Part-Time   
2           NaN   11.0           NaN          NaN        Part-Time   

  Education.Required               Created.At               Updated.At  
0                NaN  2013-03-12 02:08:28 UTC  2014-08-16 15:35:36 UTC  
1                NaN  2013-04-12 08:36:36 UTC  2014-08-16 15:35:36 UTC  
2                NaN  2013-07-16 09:34:10 UTC  2014-08-16 15:35:37 UTC  

[3 rows x 23 columns]

In [15]:
job_df['Title']

0                                     Server @ Tacolicious
1                         Kitchen Staff/Chef @ Claude Lane
2                     Bartender @ Machka Restaurants Corp.
3                                  Server @ Teriyaki House
4              Kitchen Staff/Chef @ Rosa Mexicano - Sunset
                               ...                        
84085    Book Keeper @ National Japanese American Histo...
84086                   Kitchen Staff/Chef @ Emporio Rulli
84087                                    Driver @ Onigilly
84088                 Line Cook @ Machka Restaurants Corp.
84089                           Cashier @ Kazoo Restaurant
Name: Title, Length: 84090, dtype: object

In [16]:
print(len(job_df))
job_df = job_df[['Status', 'Title', 'Position', 'Company', 'Job.Description']]
print(len(job_df))

84090
84090


The first print(len(job_df)) prints the number of rows in the original job_df DataFrame.  
The second print(len(job_df)) prints the number of rows again after selecting specific columns.  
If no rows were removed during the column selection, the output of both print(len(job_df)) calls will be the same.

In [17]:
job_df.shape

(84090, 5)

#### Total 84090 rows and 5 col found in our dataset

In [18]:
job_df['Job.Description'][1]

' \r\n\r\nNew French Brasserie in S.F. Financial District Seeks Chef\r\nWe are seeking an energetic, dynamic chef to take charge and grow with our company. Our ideal candidate is a motivated self-starter, has a great work ethic and is ready for the challenge of building their own team with executive support. \r\n\r\nThis position is a perfect fit for a talented chef ready to take the next step, someone who can multi task in a high volume kitchen and has exceptional organizational skills. Position requirements are experience with French cuisine, a minimum of 5 years as a sous chef in high volume, full service restaurant. He or she must be a team player, leading by example and working side by side with other members of the team. \r\n\r\nComputer and management skills are a must, as are experience in ordering, inventory and cost control. The right candidate will have previously demonstrated an understanding of leadership and accountability, and an infectious, energetic approach to problem

#### job_df['Job.Description'][1] will return the second job description in the DataFrame, assuming the DataFrame is indexed from 0 to N.

# NLTK

In [19]:
from nltk.corpus import stopwords
import nltk
import re
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()



#### The code imports useful tools from NLTK for text processing, such as stopwords removal, stemming, and regular expressions.
The PorterStemmer is used to reduce words to their root form, which helps simplify and standardize the text for further analysis, such as text classification or sentiment analysis.

# Data Cleaning

In [20]:
def cleaning (txt):
    cleanned_txt = re.sub(r'[^a-zA-Z0-9\s]','',txt)
    tokens=nltk.word_tokenize(cleanned_txt.lower())
    return cleanned_txt
                          

In [21]:
cleaning('this is my book with 0987 digits and \n\n\r &^%')

'this is my book with 0987 digits and \n\n\r '

In [22]:
def cleaning(txt):
    # Remove non-alphanumeric characters
    cleaned_txt = re.sub(r'[^a-zA-Z0-9\s]', '', txt)
    
    # Tokenize the text
    tokens = nltk.word_tokenize(cleaned_txt.lower())
    
    # Initialize the stemmer
    ps = PorterStemmer()

    # Remove stopwords and apply stemming
    stemmed_words = [ps.stem(word) for word in tokens if word not in stopwords.words('english')]
    
    return stemmed_words
    
    # Combine the stemmed words into a single string
    result = " ".join(stemmed_words)
    
    return result

# Test the function
result = cleaning('this is my book with 0987 digits loving moving driven and \n\n\r &^%')
print(result)

['book', '0987', 'digit', 'love', 'move', 'driven']


In [23]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer

# Downloading necessary NLTK data
nltk.download('punkt')
nltk.download('stopwords')

def cleaning(txt):
    # Remove non-alphanumeric characters
    cleaned_txt = re.sub(r'[^a-zA-Z0-9\s]', '', txt)
    
    # Tokenize the text
    tokens = nltk.word_tokenize(cleaned_txt.lower())
    
    # Initialize the stemmer
    ps = PorterStemmer()
    
    # Remove stopwords and apply stemming
    stemmed_words = [ps.stem(word) for word in tokens if word not in stopwords.words('english')]
    
    # Combine the stemmed words into a single string
    result = ' '.join(stemmed_words)  # Convert list to string
    
    return result

# Test the function
result = cleaning('this is my book with 0987 digits loving moving driven and \n\n\r &^%')
print(result)

book 0987 digit love move driven


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\vnav2\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\vnav2\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [24]:
job_df = job_df.sample(n=1000, random_state = 42)
job_df.shape

(1000, 5)

In [25]:
job_df['Title'].apply(lambda x: cleaning(x))

64119                       site director knowledg univers
35827                          administr assist officeteam
72100                     account manag chi payment system
46355    outsid wholesal sale rep parttim river front c...
34166    custom servic rep help peopl hear loss captioncal
                               ...                        
66282    sale repres sale associ entri level vector market
39515                             staff account accountemp
69231                    unarm secur offic us secur associ
69618              line cook crown plaza independ own oper
6144     kitchen manag job detail artesian hotel casino...
Name: Title, Length: 1000, dtype: object

In [26]:
job_df['Title'] = job_df['Title'].astype(str).apply(lambda x: cleaning(x))
job_df['Job.Description'] = job_df['Job.Description'].astype(str).apply(lambda x: cleaning(x))

#### The code cleans the text in the 'Title' and 'Job.Description' columns of the job_df DataFrame by:

1. Converting values to strings using .astype(str).
2. Applying a custom cleaning function to each value using .apply(lambda x: cleaning(x)).
This process ensures both columns are preprocessed uniformly, typically for tasks like removing punctuation, lowercasing, or other text cleaning operations as defined in the cleaning function.








In [27]:
job_df['Title']

64119                       site director knowledg univers
35827                          administr assist officeteam
72100                     account manag chi payment system
46355    outsid wholesal sale rep parttim river front c...
34166    custom servic rep help peopl hear loss captioncal
                               ...                        
66282    sale repres sale associ entri level vector market
39515                             staff account accountemp
69231                    unarm secur offic us secur associ
69618              line cook crown plaza independ own oper
6144     kitchen manag job detail artesian hotel casino...
Name: Title, Length: 1000, dtype: object

In [28]:
job_df['Job.Description'][64119]

'job summari knowledg univers ku site director site leader inspir children teacher alik learn grow passion educ excel confid teach children adult use nation recogn curriculum framework creat uniqu engag classroom experi commit make site success know meaning relationship children famili team import success fulli engag enthusiast work eager share knowledg other job respons essenti function basic expect site director cours creativ new way meet exceed expect encourag long requir essenti function also met supervis children staff record keep licens record child file lesson plan implement mainten safe welcom classroom environ build relationship commun school recruit new student program applic must strong organiz skill'

In [29]:
job_df['new_col'] = job_df['Title'] + " " + job_df['Job.Description']

In [30]:
job_df.head(3)

Status                             Title                  Position  \
64119   open    site director knowledg univers             Site Director   
35827   open       administr assist officeteam  Administrative Assistant   
72100   open  account manag chi payment system           Account Manager   

                   Company                                    Job.Description  \
64119   Knowledge Universe  job summari knowledg univers ku site director ...   
35827           OfficeTeam  ref id 03110118480classif secretaryadmin asstc...   
72100  CHI Payment Systems  yoursquor energet motiv hardwork look prosper ...   

                                                 new_col  
64119  site director knowledg univers job summari kno...  
35827  administr assist officeteam ref id 03110118480...  
72100  account manag chi payment system yoursquor ene...

# Natural Language Processing

# vectorization

#### TfidfVectorizer: Converts text into numerical vectors using TF-IDF, which is useful for processing text data for machine learning or NLP.
cosine_similarity: Measures the similarity between text documents (or their vector representations) based on their content, useful for identifying related documents or clustering similar items.

In [31]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [32]:
tfidf = TfidfVectorizer(stop_words='english')
matrix = tfidf.fit_transform(job_df['new_col'])
similarity = cosine_similarity(matrix)

#### Step 1: Convert the text data from the new_col column of job_df into numerical vectors using the TF-IDF method.
#### Step 2: Compute the cosine similarity between each pair of documents in the new_col column to measure how similar they are.

In [33]:
similarity

array([[1.        , 0.04466151, 0.02193005, ..., 0.05658977, 0.02584512,
        0.08124254],
       [0.04466151, 1.        , 0.02996055, ..., 0.03590061, 0.00568821,
        0.0259882 ],
       [0.02193005, 0.02996055, 1.        , ..., 0.05656006, 0.02807874,
        0.04050679],
       ...,
       [0.05658977, 0.03590061, 0.05656006, ..., 1.        , 0.06369181,
        0.11336692],
       [0.02584512, 0.00568821, 0.02807874, ..., 0.06369181, 1.        ,
        0.38689315],
       [0.08124254, 0.0259882 , 0.04050679, ..., 0.11336692, 0.38689315,
        1.        ]])

In [34]:
enumerate(similarity[0])

# Recommendation System

In [35]:
def recommend(title):

        indx = job_df[job_df['Title'] == title].index[0]
        indx = job_df.index.get_loc(indx)
        distances = sorted(list(enumerate(similarity[indx])), key=lambda x: x[1], reverse=True)[1:20]

        jobs = []
        for i in distances:
            jobs.append(job_df.iloc[i[0]].Title)
        return jobs
        

# This code defines a function `recommend(title)` that provides job recommendations based on a given job title. Here's an explanation of each part of the code:

 1. **`indx = job_df[job_df['Title'] == title].index[0]`**:
   - This line finds the index of the job with the given `title` in the DataFrame `job_df` (assumed to be a pandas DataFrame containing job information, including a 'Title' column).
   - `job_df['Title'] == title` creates a boolean Series where the value is `True` if the job title matches the `title` parameter.
   - `.index[0]` retrieves the index of the first match.

2. **`indx = job_df.index.get_loc(indx)`**:
   - This converts the index `indx` into a positional index, which may be useful if the DataFrame is not ordered sequentially.

3. **`distances = sorted(list(enumerate(similarity[indx])), key=lambda x: x[1], reverse=True)[1:20]`**:
   - `similarity[indx]` is assumed to be a list or array representing the similarity scores of all job titles relative to the job at `indx`.
   - `enumerate(similarity[indx])` creates an iterable of index-value pairs from the `similarity` list/array, where each pair represents the index of another job and its similarity score to the job at `indx`.
   - `sorted(..., key=lambda x: x[1], reverse=True)` sorts these pairs by the similarity score (second element of the tuple) in descending order (so the most similar jobs come first).
   - `[1:20]` slices the sorted list to exclude the first item (which is the job itself) and keeps the top 19 most similar jobs.

4. **`jobs = []`**:
   - Initializes an empty list to store the recommended job titles.

5. **`for i in distances:`**:
   - Loops through the top 19 most similar jobs.

6. **`jobs.append(job_df.iloc[i[0]].Title)`**:
   - For each item in `distances`, `i[0]` represents the index of a recommended job.
   - `job_df.iloc[i[0]]` retrieves the job row from `job_df`, and `.Title` extracts the job title.
   - The job title is appended to the `jobs` list.

7. **`return jobs`**:
   - Returns the list of recommended job titles.

In summary, this function takes a job title as input, finds the most similar jobs from a precomputed similarity matrix, and returns the titles of the top 19 most similar jobs.

In [36]:
recommend('site director knowledg univers')

['site director knowledg univers educ llc',
 'teacher knowledg univers',
 'assist teacher knowledg univers',
 'assist teacher knowledg univers',
 'cook knowledg univers',
 'immedi open assist teacher la petit academi',
 'summer school age assist children courtyard',
 'hvac instructor vatterott educ center',
 'temporari coordin site oper macyscom maci',
 'medic offic assist instructor concord career colleg inc',
 'fellowship program coordin connecticut children medic center',
 'pharmaci adjunct instructor brown macki colleg',
 'youth camp residenti assist activ coordin new york ny el educ servic',
 'design bed sell specialist full timepart time bloomingdal mall short hill nj maci',
 'secur offic 100000 job coalit securita usa',
 'secur offic regular securita usa',
 'handbag sell specialist part time bloomingdal chevi chase md bloomingdal',
 'men design sell specialist full timepart time bloomingdal newport beach ca bloomingdal',
 'women shoe sale profession part time bloomingdal 59th st

In [37]:
import pickle
pickle.dump(job_df,open('job_df.pkl','wb'))
pickle.dump(similarity,open('similarity.pkl','wb'))

In [ ]:
#The code saves two objects (job_df and similarity) to disk as pickle files. 
#These files (job_df.pkl and similarity.pkl) can later be loaded back into memory using pickle.load(). 
#This process is useful for persisting data that you want to use later without needing to recompute or reload it each time.

In [38]:
import os
print(os.getcwd())


C:\Users\vnav2


In [39]:
import pandas as pd
print(pd.__version__)


2.1.4


In [ ]:
# One can run same code in Py charm for better result.

In [40]:
!pip install streamlit


In [41]:
# Streamlit app
st.title('Job Recommendation System')

# Job search - select job title
selected_title = st.selectbox('Search job by title', job_df['Title'])

# Display job recommendations
recommended_jobs = recommendation(selected_title)

if isinstance(recommended_jobs, pd.DataFrame):
    st.write("Recommended Jobs:")
    st.write(recommended_jobs)
else:
    st.write(recommended_jobs)

In [45]:
import subprocess

# Run the Streamlit app in a subprocess
subprocess.run(["streamlit", "run", "app.py"])


CompletedProcess(args=['streamlit', 'run', 'app.py'], returncode=1)

In [46]:
%%writefile app.py
import streamlit as st
import pandas as pd

# Load your dataset (you can modify this to your dataset path)
job_df = pd.read_csv(r"C:\Users\vnav2\OneDrive\Desktop\Combined_Jobs_Final.csv")  # replace 'your_dataset.csv' with your dataset path

st.title('Job Recommendation System')

# Job search - select job title
selected_title = st.selectbox('Search job by title', job_df['Title'].unique())

def recommendation(title):
    recommended_jobs = job_df[job_df['Title'].str.contains(title, case=False, na=False)]
    if not recommended_jobs.empty:
        return recommended_jobs
    else:
        return "No job recommendations found."

# Display job recommendations
recommended_jobs = recommendation(selected_title)

if isinstance(recommended_jobs, pd.DataFrame):
    st.write("Recommended Jobs:")
    st.write(recommended_jobs[['Title', 'Company', 'Location', 'Salary']])
else:
    st.write(recommended_jobs)


Writing app.py


In [48]:
import subprocess
import time
import webbrowser

# Run Streamlit app in subprocess
subprocess.Popen(["streamlit", "run", "app.py"])

# Wait for Streamlit to start
time.sleep(5)

# Open the app in the default web browser
webbrowser.open("http://localhost:8501")


True

In [57]:
!streamlit run "C:/Users/vnav2/PycharmProjects/pythonProject1/Job Recommendation system app.py"

# streamlit run "C:\Users\vnav2\PycharmProjects\pythonProject1\Job Recommendation system app.py
# use above link in py charm terminal and you can see result

Traceback (most recent call last):
  File "C:\Users\vnav2\anaconda3\Scripts\streamlit-script.py", line 6, in <module>
    from streamlit.cli import main
ModuleNotFoundError: No module named 'streamlit.cli'
